### DISCUSSION TITLE
Two way compiled binding in "Run" element causing InvalidCastException when "StringFormat" is set

### DISCUSSION THREAD

**Discussion by stewpend0us** [AvaloniaUI/Avalonia#17323] [Q&A]
### Describe the bug

I have a ```TextBlock``` containing several ```Run```s. The first ```Run``` ```Text``` is binding a ```double``` and the ```StringFormat``` is set. I get the following exception unless I do one of:

* Set ```Mode=OneWay```
* Replace the ```Run``` with ```TextBlock```
* Remove the ```StringFormat```

Exception:
```
System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Double'.
   at CompiledAvaloniaXaml.XamlIlHelpers.toolkit.avalonia.ViewModels.MyViewModel,toolkit.avalonia.MyDouble!Setter(Object, Object)
   at Avalonia.Data.Core.ClrPropertyInfo.Set(Object target, Object value)
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.SetValue(Object value, BindingPriority priority)
   at Avalonia.Data.Core.ExpressionNodes.PropertyAccessorNode.WriteValueToSource(Object value, IReadOnlyList`1 nodes)
   at Avalonia.Data.Core.BindingExpression.WriteValueToSource(Object value)
```

### To Reproduce

I'm not sure how re-producible this will be? My colleague does not have this issue in the same code base. We're both using the Rider IDE on Ubuntu 22.04. Compiled bindings are turned on by default.

This throws the exception:
```
...
<TextBlock Grid.Row="1" Grid.Column="1" Background="{DynamicResource OverlayBackground}">
  <Run Text="{Binding MyDouble, StringFormat='MyDouble: {0:N1} ', TargetNullValue='MyDouble: ***' , FallbackValue='error' }"/>
  <Run Text="{Binding Units, TargetNullValue=' units', FallbackValue='error' }" />
</TextBlock>
...
```
This does not:
```
...
<TextBlock Grid.Row="1" Grid.Column="1" Background="{DynamicResource OverlayBackground}">
  <Run Text="{Binding MyDouble, Mode=OneWay, StringFormat='MyDouble: {0:N1} ', TargetNullValue='MyDouble: ***' , FallbackValue='error' }"/>
  <Run Text="{Binding Units, TargetNullValue=' units', FallbackValue='error' }" />
</TextBlock>
...
```
This does not:
```
...
<TextBlock Grid.Row="1" Grid.Column="1" Background="{DynamicResource OverlayBackground}">
  <TextBlock Text="{Binding MyDouble, StringFormat='MyDouble: {0:N1} ', TargetNullValue='MyDouble: ***' , FallbackValue='error' }"/>
  <Run Text="{Binding Units, TargetNullValue=' units', FallbackValue='error' }" />
</TextBlock>
...
```
This does not:
```
...
<TextBlock Grid.Row="1" Grid.Column="1" Background="{DynamicResource OverlayBackground}">
  <Run Text="{Binding MyDouble, TargetNullValue='MyDouble: ***' , FallbackValue='error' }"/>
  <Run Text="{Binding Units, TargetNullValue=' units', FallbackValue='error' }" />
</TextBlock>
...
```

### Expected behavior

Odd that this seems to be specific to the ```Run``` tag with ```StringFormat``` set? Is the default mode not ```OneWay``` in that case? Why is it different than using another ```TextBlock```?

### Avalonia version

11.1.3 and 11.1.4

### OS

Linux

### Additional context

we're also using CommunityToolkit 8.3.2 if that matters.

**Comment by timunie** [comment]
That's by design if the BindingMode is TwoWay. However it shouldn't be TwoWay for a Run, as the user can't manipulate it either 🤔 

Can you please attach a minimal sample to reproduce?

**Comment by stewpend0us** [comment]
Sure what do you need? axaml and view model? project file? solution file?

**Comment by timunie** [comment]
A zip of a minimal project without obj and bin folders. But all xaml and viewmoel / converter etc

**Comment by stewpend0us** [comment]
Alright. I wasn't able to reproduce this. Made a small example and couldn't get it to throw. Also tried the original repo on a different computer and it all seems normal. I'll try again in the original repo and original computer tomorrow...but like I said my co-worker didn't have this issue either. Must be something specific to that machine?

**Comment by timunie** [comment]
I have no idea, but it's really strange. Will move this into Q&A, we can make it an issue in case you have a way to reproduce.

**Comment by stewpend0us** [comment]
Dug a little more on the machine with the issue. I think I may have failed to launch it in such a way to reproduce the issue at home...I'll try again tonight.

That being said. The exception is happening in "CastHelpers.cs" "Unbox" method (Specifically CastHelpers.Unbox_Helper()). In that function the 'obj' argument is always what I have set for the FallbackValue. It appears to be upset that it can't cast obj (string) to a Double. If I remove the FallbackValue the exception goes away.

**Comment by stewpend0us** [comment]
@timunie I was able to reproduce it! Here are the files.

Ultimately it's trying to cast a string to a double somewhere along the line.

[reproduce_bug.zip](https://github.com/user-attachments/files/17484644/reproduce_bug.zip)

**Reply by stewpend0us** [reply]
![image](https://github.com/user-attachments/assets/3c192f9a-f4fd-4c35-8fdd-c7b1d409ee0c)

**Reply by timunie** [reply]
For some reason your sample doesn't crash for me. However, you may want to set  Binding-Mode to `OneWay` in order to solve your specific use-case.

> [!NOTE]
> Afaik it is by design, that `Run.Text` produces a two-way binding. This binding cannot covert back a new value due to the formatted number.

**Reply by stewpend0us** [reply]
Ok. I think the real answer is we just don't need a fallback. It all 'works' but in debug mode I get the 'User-unhandled exception' pop up.
